### PR TITLE
Reviewed : 1209-freeze-input-fields-based-on-conditions

### PR DESCRIPTION
**Pull request by A4family**
… activity level or in transaction level according to conditions

### CODE REVIEW

File: app/Http/Controllers/Admin/Activity/ActivityController.php [hunk 1]
```diff
@@ -198,6 +198,48 @@ public function show($id): View|JsonResponse|RedirectResponse
         }
     }
 
+    /**
+     * @throws \JsonException
```
Comment (sriza) [inline app/Http/Controllers/Admin/Activity/ActivityController.php:202]: method definition missing

File: app/Http/Controllers/Admin/Activity/ActivityController.php [hunk 2]
```diff
@@ -198,6 +198,48 @@ public function show($id): View|JsonResponse|RedirectResponse
...
+     */
+    public function getElementJsonSchema($activity): array
+    {
+        $element = readElementJsonSchema();
+        $hasDefinedInTransaction = $this->transactionService->hasRecipientRegionOrCountryDefinedInTransaction($activity->id);
+
```
Comment (sriza) [inline app/Http/Controllers/Admin/Activity/ActivityController.php:208]: no need for empty line here

File: app/Http/Controllers/Admin/Activity/ActivityController.php [hunk 3]
```diff
@@ -198,6 +198,48 @@ public function show($id): View|JsonResponse|RedirectResponse
...
+        }
+
+        $warning_info_text = match (true) {
+            $emptyRecipientRegionOrCountryTransactionCount > 0 && $emptySectorTransactionCount > 0 => 'Recipient Region , Recipient Country and Sector are declared at transaction level. Some of the transactions are missing these values.',
+            $emptyRecipientRegionOrCountryTransactionCount > 0 && $emptySectorTransactionCount === 0 => 'Recipient Region and Recipient Country is declared at transaction level. Some of the transactions are missing these values.',
+            $emptySectorTransactionCount > 0 && $emptyRecipientRegionOrCountryTransactionCount === 0  => 'Sector is declared at transaction level. Some of the transactions are missing this value.',
+            default => ''
+        };
+
+        $element['transactions']['warning_info_text'] = $warning_info_text;
```
Comment (sriza) [inline app/Http/Controllers/Admin/Activity/ActivityController.php:238]: we can assign warning text directly to element->transaction->warning_info_text

File: app/Http/Controllers/Admin/Activity/RecipientRegionController.php [hunk 4]
```diff
@@ -76,4 +77,31 @@ public function update(RecipientRegionRequest $request, $id): JsonResponse|Redir
...
+     * append freeze and info_text if recipient country or region exists in any one of the activity transactions
+     * if exists then it freezes the section.
+     *
+     * @param $activity
+     *
+     * @throws JsonException
+     *
+     * @return array
+     */
+    public function getRecipientRegionManipulatedElementSchema($activity): array
```
Comment (sriza) [inline app/Http/Controllers/Admin/Activity/RecipientRegionController.php:91]: this code can be reused for both recipient region and country. considering adding the test to activity service

File: app/Http/Controllers/Admin/Activity/TransactionController.php [hunk 5]
```diff
@@ -216,6 +217,108 @@ public function edit($activityId, $transactionId): Factory|View|RedirectResponse
...
+    /**
+     * appends warning info text in recipient region or country in transaction level.
+     *
+     * @param $activity
+     * @param $element
+     * @param $transactionId
+     *
+     * @return void
+     */
+    public function appendInfoTextForRecipientRegionAndCountryInTransaction($activity, &$element, $transactionId): void
```
Comment (sriza) [inline app/Http/Controllers/Admin/Activity/TransactionController.php:263]: place this code in transaction service

File: app/Http/Controllers/Admin/Activity/TransactionController.php [hunk 6]
```diff
@@ -216,6 +217,108 @@ public function edit($activityId, $transactionId): Factory|View|RedirectResponse
...
+    /**
+     * Adds warning info text in sector in transaction level.
+     *
+     * @param $activity
+     * @param $element
+     * @param $transactionId
+     *
+     * @return void
+     */
+    public function appendInfoTextForSectorInTransaction($activity, &$element, $transactionId): void
```
Comment (sriza) [inline app/Http/Controllers/Admin/Activity/TransactionController.php:298]: place this code in the transaction service

File: app/IATI/Services/Activity/RecipientRegionService.php [hunk 7]
```diff
@@ -71,14 +86,16 @@ public function getActivityData($id): object
      * @param $activityRecipientRegion
      *
      * @return bool
```
Comment (sriza) [inline app/IATI/Services/Activity/RecipientRegionService.php:88]: place throw before return and add an empty line inbetween

File: app/IATI/Services/Activity/RecipientRegionService.php [hunk 8]
```diff
@@ -166,19 +182,21 @@ public function sanitizeRecipientRegionData($activityRecipientRegion): array
      *
      * @param int $id
      * @param array $data
-     * @param int $totalRecipientRegionPercentage
+     * @param $totalRecipientRegionPercentage
+     *
      * @return array
+     *
+     * @throws BindingResolutionException
+     * @throws JsonException
```
Comment (sriza) [inline app/IATI/Services/Activity/RecipientRegionService.php:190]: place throw before return

File: app/IATI/Services/ElementCompleteService.php [hunk 9]
```diff
@@ -403,7 +406,7 @@ public function isPolicyMarkerElementCompleted($activity): bool
      * @param $activity
      *
      * @return bool
-     * @throws \JsonException
+     * @throws JsonException
```
Comment (sriza) [inline app/IATI/Services/ElementCompleteService.php:409]: place all the throws before return and leave an empty line between them in this file

File: app/Observers/TransactionObserver.php [hunk 10]
```diff
@@ -193,4 +196,14 @@ public function deleted(Transaction $transaction): void
         $activityObj->element_status = $elementStatus;
         $activityObj->save();
     }
+
+    private function recipientRegionIsEmptyInTransaction(mixed $recipientRegion): bool
```
Comment (sriza) [inline app/Observers/TransactionObserver.php:200]: add php docs

File: app/Observers/TransactionObserver.php [hunk 11]
```diff
@@ -193,4 +196,14 @@ public function deleted(Transaction $transaction): void
...
+    {
+        return (bool) Arr::get($recipientRegion, '0.region_vocabulary', false);
+    }
+
+    private function recipientCountryIsEmptyInTransaction(mixed $recipientCountry): bool
```
Comment (sriza) [inline app/Observers/TransactionObserver.php:205]: add php docs

File: resources/views/vendor/laravel-form-builder/collection.php [hunk 12]
```diff
@@ -36,20 +36,54 @@ class="<?= strtolower($options['label']) === "narrative" ? $options['options']['
...
+
+            if(isset($options['options']['warning_info_text']) && !empty($options['options']['warning_info_text']))
+            {
+                $errorSection = '<div class="mt-2 flex items-center bg-eggshell pt-2 pr-4 pb-2 pl-4 text-xs rounded-md font-normal">
+                                    <svg
+                                        class="elements-svg"
+                                        width="18"
+                                        height="18"
+                                        viewBox="0 0 18 18"
+                                        fill="none"
```
Comment (sriza) [inline resources/views/vendor/laravel-form-builder/collection.php:58]: add svg to different file then use it as such: <svg-vue icon="svg name" class="mr-2"></svg-vue>

File: resources/views/vendor/laravel-form-builder/collection.php [hunk 13]
```diff
@@ -36,20 +36,54 @@ class="<?= strtolower($options['label']) === "narrative" ? $options['options']['
...
+                                        />
+                                    </svg>
+                                    <div>'.$options['options']['warning_info_text'].'</div>
+                                </div>';
+            }
 
             if ($showError && isset($errors) && $errors->hasBag($errorBag)) {
                 foreach ($errors->getBag($errorBag)->get($nameKey) as $err) {
-                    $error = $error . '<div class="text-danger-error">' . $err . '</div>';
+                    if($err !== "")
```
Comment (sriza) [inline resources/views/vendor/laravel-form-builder/collection.php:72]: !empty($err)

**Review by Sanilblank — CHANGES_REQUESTED** [review]

**Review by sriza — APPROVED** [review]

**Review by Sanilblank — APPROVED** [review]

**Review by Sanilblank — APPROVED** [review]

**Pull request merged**